Feature/Add plugin system

Blokje5 opened the pull request:
Fixes #226 

This PR adds a plugin system inspired by the helm plugin system.

A plugin is described by a plugin.yaml file. For example:

```yaml
name: "kubectl"
version: "0.1.0"
usage: conftest kubectl (TYPE[.VERSION][.GROUP] [NAME] | TYPE[.VERSION][.GROUP]/NAME).
description: |-
  A Kubectl plugin for using Conftest to test objects in Kubernetes using Open Policy Agent.
  Usage: conftest kubectl (TYPE[.VERSION][.GROUP] [NAME] | TYPE[.VERSION][.GROUP]/NAME).
command: $CONFTEST_PLUGIN_DIR/kubectl-conftest.sh
```

The $CONFTEST_PLUGIN_DIR is a special environment variable made available for the plugin to find the location of the executable located alongside the `plugin.yaml` on disk.

This example plugin.yaml makes a `conftest kubectl` command available that can be used as following:

```console
conftest kubectl deployment <deployment-id> --policy examples/kubernetes/policy
```
The kubectl conftest plugin is essentially the plugin @garethr created a while ago with some small modifications.

Plugins can be downloaded with the `conftest plugin install` command, which uses go-getter to download the plugin and place it in a caching folder (`~/.conftest/plugins`). The full url is used as the name of the directory in the caching folder. E.g. `https-github-com-instrumenta-conftest-example-plugin-kubectl`.

Initally I was going in the direction of allowing plugins to be run at specific points in the testing cycle (e.g. pre-parse). This would mean a plugin command could be as simple as `kubectl $1 $2 -o json`. We could then deal with ensuring every argument is passed correctly in conftest and ensure the output of the command is correctly piped in the conftest test command. But this would lead to a lot of additional complexity on the conftest side + it would mean plugins don't have a lot of flexibility in how they behave as they can only run in certain parts of the testing cycle. Hence I decided to go with the helm plugin approach.

**NOTE**: We should squash merge this PR.

boranx commented on internal/commands/load_plugins.go:
```diff
@@ -0,0 +1,37 @@
+import (
+	"context"
+	"fmt"
+	"github.com/instrumenta/conftest/plugin"
```
new line

boranx commented on plugin/plugin.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+// SetEnv configures the available environment for the plugin
+func (p *Plugin) SetEnv(env []string) *Plugin {
```
looks like it's created but not used 
what's the rationale of creating this?

Blokje5 replied:
it is indeed not need, I went for a different direction with the plugin system initially.

boranx commented on plugin/plugin.go:
```diff
@@ -0,0 +1,210 @@
+		return nil, fmt.Errorf("read plugin cache directory: %v", err)
+	}
+
+	// TODO log here instead? Otherwise conftest breaks on faulty plugins in the cache
```
Imho, it should stay as it is (break when we face with any faulty on the plugin)

boranx commented on internal/commands/load_plugins.go:
```diff
@@ -0,0 +1,37 @@
+func loadPlugins(ctx context.Context) ([]*cobra.Command, error) {
+	plugins, err := plugin.FindPlugins()
+	if err != nil {
+		return nil, fmt.Errorf("loading plugins: %v", err)
```
what's the rationale of not using `%w` pattern just like we applied before to the rest of the code?

Blokje5 replied:
My bad, I default to `%v` so I just missed it.

Blokje5 commented:
@boranx Thanks for reviewing. I updated the code based on your comments and added a simple acceptance test. I also squashed everything into a single commit. Please let me know if you have additional comments.

garethr commented on examples/plugins/kubectl/plugin.yaml:
```diff
@@ -0,0 +1,7 @@
+version: "0.1.0"
+usage: conftest kubectl (TYPE[.VERSION][.GROUP] [NAME] | TYPE[.VERSION][.GROUP]/NAME).
+description: |-
+  A Kubectl plugin for using Conftest to test objects in Kubernetes using Open Policy Agent.
```
This is actually now a conftest plugin for using kubectl :)

Blokje5 replied:
Haha good catch!

Blokje5 commented:
@garethr I have thought about adding install hooks. In the end I decided the keep the plugin system to a minimum for now, and indeed focus on iterating once we have a bit of an idea of usage. I can see the hooks being part of conftest after a while as well.
